Improve `txout` listing and balance APIs for redesigned structures

evanlinjin opened the pull request:
### Description

As noted in https://github.com/bitcoindevkit/bdk/issues/971#issuecomment-1542408941.

Instead of relying on a `OwnedIndexer` trait to filter for relevant txouts, we move the listing and balance methods from `IndexedTxGraph` to `TxGraph` and add an additional input (list of relevant outpoints) to these methods.

This provides a simpler implementation and a more flexible API.

#### Other Fixes

The `Persist::commit` method is fixed in 96b10751325dc1bced09f9e864438216400e4ba3.

Previously, regardless of whether writing to persistence backend is successful or not, the logic always cleared `self.staged`. This is changed to only clear `self.staged` after successful write.

Additionally, the written changeset (if any) is returned, and `PersistBackend::write_changes` will not be called if `self.staged` is empty.

### Notes to the reviewers

Yes, slightly more boilerplate to do the same things, but less code to maintain and a much more flexible API. Very worth it IMO.

### Changelog notice

### Checklists

#### All Submissions:

* [x] I've signed all my commits
* [x] I followed the [contribution guidelines](https://github.com/bitcoindevkit/bdk/blob/master/CONTRIBUTING.md)
* [x] I ran `cargo fmt` and `cargo clippy` before committing

#### New Features:

* [x] I've added tests for the new feature
* [x] I've added docs for the new feature

evanlinjin committed:
Improve `txout` listing and balance APIs for redesigned structures

Instead of relying on a `OwnedIndexer` trait to filter for relevant
txouts, we move the listing and balance methods from `IndexedTxGraph` to
`TxGraph` and add an additional input (list of relevant outpoints) to
these methods.

This provides a simpler implementation and a more flexible API.

LLFourn commented on crates/chain/src/indexed_tx_graph.rs:
```diff
@@ -29,6 +26,14 @@ impl<A, I: Default> Default for IndexedTxGraph<A, I> {
 
 impl<A, I> IndexedTxGraph<A, I> {
+    /// Construct a new [`IndexedTxGraph`] with a given `index`.
+    pub fn new(index: I) -> Self {
```
Why can't you pass in both graph and index here. Just index everything in the graph.

evanlinjin replied:
Indexing data may result in `I::Additions`. Then the method needs to return both `Self` and `IndexedAdditions`. I don't think this is the right API.

The intention for constructing `IndexedTxGraph` with only an index is so you can set up things like keychains and lookahead settings. I don't think you would ever want to add transactions into `TxGraph` before this stage.

LLFourn replied:
Ok understood. I think there should be some way of taking an existing graph and and index and creating an indexed tx graph but we can leave that for another time.

LLFourn commented on crates/chain/src/tx_graph.rs:
```diff
@@ -762,107 +763,190 @@ impl<A: Anchor> TxGraph<A> {
+        outpoints: impl IntoIterator<Item = (S, OutPoint)> + 'a,
+    ) -> impl Iterator<Item = Result<(S, FullTxOut<ObservedAs<A>>), C::Error>> + 'a {
+        self.try_filter_chain_txouts(chain, chain_tip, outpoints)
+            .filter(|r| !matches!(r, Ok((_, full_txo)) if full_txo.spent_by.is_some()))
```
no.

evanlinjin replied:
Clippy will complain otherwise 🫤

evanlinjin replied:
Jokes, there is a way of doing it that improves readability by alot.

evanlinjin replied:
https://github.com/bitcoindevkit/bdk/blob/fb75aa94a9f47d53ce7246c14d739fc53dc38197/crates/chain/src/tx_graph.rs#L867-L873

LLFourn commented on crates/chain/src/tx_graph.rs:
```diff
@@ -762,107 +763,190 @@ impl<A: Anchor> TxGraph<A> {
     ///
-    /// Floating ouputs are not iterated over.
+    /// `outpoints` is a list of outpoints we are interested in, coupled with the associated txout's
+    /// script pubkey index (`S`).
```
why does this have to be a script pubkey index. It's just any data you want to associate with the outpoint. You should suggest if you don't have a natural choice suggest to the reader they can just do `vec_of_outpoints.iter().enumerate()` for example.

The type parameter should probably be named `I` and emphasize that it's there for convenience  and can be `()` if you just don't want one. It would be nice if you could even do without it but it looks like necessary plumbing to make getting the balance easy.

evanlinjin replied:
~How about instead of `I`, we have:~

```rust
impl IntoIterator<Item = impl AsRef<OutPoint>>
```

Okay, bad idea. There is a lot of friction in getting things to implement `AsRef`.

evanlinjin replied:
Changed generic bound to `OI` (outpoint identifier) and improved docs: ed89de752cf20cb80deefda3ddf65a137b669288

evanlinjin committed:
Fix and improve `Persist::commit` method

Previously, regardless of whether writing to persistence backend is
successful or not, the logic always cleared `self.staged`. This is
changed to only clear `self.staged` after successful write.

Additionally, the written changeset (if any) is returned, and
`PersistBackend::write_changes` will not be called if `self.staged` is
empty.

evanlinjin committed:
Clarify `TxGraph::try_filter_chain_unspents` logic

evanlinjin committed:
Improve txout filter/listing method docs for `TxGraph`

Rename the `S` trait bound to `OI` (outpoint index) to emphasize it's
not only for spk indexing.

LLFourn reviewed (approved):
I think this is better. Looking forward to seeing how this works in the examples.

LLFourn commented:
ACK ed89de752cf20cb80deefda3ddf65a137b669288
